// @robinsage/robinbase/services/api/apiApp.js
(function(){

    var Debug = require('../../helpers/Debug.js');
    var Config = require('../../config.js');
    var express = require('express');
    var bodyParser = require('body-parser');
    var renderFunctions = require('../../helpers/RenderFunctions.js');
    var qs = require('qs');
    const {STATUS_CODES} = require('http');

    //include models...
    //var Account = Config.apiModels['accounts'];
   // var Token = Config.apiModels['tokens'];

    var ApiApp = function ApiApp(port = Config.RB_API_PORT)
    {
        var self = this;

        self.init = function(_routes)
        {
            var dbApp = express();
            dbApp.disable('x-powered-by');

            dbApp.set('query parser', function(query)
            {
                return qs.parse(query, {
                    depth: 100,
                });
            });

            dbApp.use(bodyParser.urlencoded({ extended: true })); // parse application/x-www-form-urlencoded
            dbApp.use(bodyParser.json());

            dbApp.use(renderFunctions);




            const routes = _routes || [require('../../routes/adminRoutes')];

            var defaultProcessData = function defaultProcessData(req, processData)
            {
                if ((typeof processData == 'undefined') || (processData == null))
                {
                    processData = {};
                }
                processData.session = req.session;
                processData.query = req.query;
                processData.body = req.body;
                processData.params = req.params;

                return processData;
            };
            dbApp.use(function(req,res,next)
            {
                res.locals.apiMode = true;
                /* if (parseInt(req.query.verbose) == 1)
                 {
                 res.locals.verbose = 1;
                 }*/
                next();
            });
            dbApp.use(function(req,res,next)
            {
                req.defaultProcessData = function(processData)
                {
                    return defaultProcessData(req,processData);
                };
                next();
            });
            routes.forEach( function (router)
            {
                dbApp.use(router);
            });

            var dbServer = dbApp.listen(port, function()
            {
                Debug.log('SERVER STARTED', 'DB Server listening on ' + Config.RB_API_PORT);
            });


            dbApp.use((err, req, res, next) =>
            {
                res.status(400).json({
                    code: err.statusCode || 500,
                    status: STATUS_CODES[err.statusCode || 500] || 'An error occurred',
                    error: err.message || err,
                });
            });

            return dbServer;

        };
    };

    module.exports = ApiApp;

}).call(this);
