// @robinsage/rb-mongo/service/db/Mongo.js
(function(){

    var Debug = require_robinbase('Debug').prefix("Mongo Connection");
    var MongoClient = require('mongodb').MongoClient;
    var MongoCursor = require('mongodb').Cursor;
    var ObjectID = require('mongodb').ObjectID;

// Connection URL

// Use connect method to connect to the Server
    var Mongo = function Mongo()
    {
        var self = this;
        self.client = null;
        self.queue = [];
        self.connect = function connect(url, callback)
        {
            var connectOptions = {domainsEnabled:true};

            MongoClient.connect(url, connectOptions, function(err, client) {
                if (err != null)
                {
                    Debug.error("Could not connect to the server at "+url);
                    return callback(err, client);

                }

                Debug.log("Connected correctly to server at "+url);
                Debug.log('attempt to switch to database');
                var dbMatches = url.match(/^mongodb:\/\/[^\/]+\/([^\/]+)$/);
                if (!Array.isArray(dbMatches))
                {
                    Debug.error("Could not select database because we could not parse the database from the url: "+url);
                    return callback(err, client);
                }
                if (dbMatches.length < 2)
                {
                    Debug.error("Could not find the database argument in the connection url: "+url);
                    return callback(err, client);
                }
                //let dbOut = client.db(dbMatches[1]);
                self.client = dbOut = client.db(dbMatches[1]);

                for (var i=0; i<self.queue.length; i++)
                {
                    self.go.apply(self, self.queue[i]);
                }
                self.queue = [];

                callback(err, dbOut);
                //db.close();
            });

            return self;
        }

        self.go = function go()
        {
            if (self.client == null)
            {
                self.queue.push(arguments);
                Debug.log("The client is not connected.  Perhaps it is not ready.");
                return new mError('mongo client is not ready to process commands');//new MongoCursor(null, null, 'null', {}, {}, {});//.CommandCursor();
            }
            var collectionStr = [].shift.apply(arguments);
            var method = [].shift.apply(arguments);
           // Debug.log("Mongo collection call", collectionStr);
           // Debug.log("Mongo method call", method);
            try
            {
                var collection = self.client.collection(collectionStr);

                return collection[method].apply(collection, arguments);
                //collection.find({}).toArray(arguments[arguments.length - 1]);
            }
            catch(e)
            {
                Debug.error("Mongo Go", e);
                return new mError(e);
            }
            return new mError('unknown issue running mon.go');
        }

        self.command = function command()
        {
            if (self.client == null)
            {
                self.queue.push(arguments);
                Debug.log("The client is not connected.  Perhaps it is not ready.");
                return new mError('mongo client is not ready to process commands');//new MongoCursor(null, null, 'null', {}, {}, {});//.CommandCursor();
            }
            try
            {
                var admin = self.client.admin();
                return admin.command.apply(admin, arguments);
            }
            catch(e)
            {
                Debug.error("Mongo Command", e);
                return new mError(e);
            }
            return new mError('unknown issue running mon.command');
        }

    }

    Mongo.ObjectID = ObjectID;

    var mError = function mError(errMsg)
    {
        this.addCursorFlag =
        this.addQueryModifier =
        this.batchSize =
        this.clone =
        this.close =
        this.comment =
        this.count =
        this.each =
        this.explain =
        this.filter =
        this.forEach =
        this.hasNext =
        this.hint =
        this.isClosed =
        this.limit =
        this.map =
        this.max =
        this.maxAwaitTimeMS =
        this.maxScan =
        this.maxTimeMS =
        this.min =
        this.next =
        this.nextObject =
        this.pause =
        this.pipe =
        this.project =
        this.read =
        this.resume =
        this.returnKey =
        this.rewind =
        this.setCursorOption =
        this.setEncoding =
        this.setReadPreference =
        this.showRecordId =
        this.skip =
        this.snapshot =
        this.sort =
        this.stream =
        this.toArray =
        this.unpipe =
        this.unshift =
        this.wrap = function(cb) { cb(errMsg, null); };
    }


    module.exports = Mongo;

}).call(this);
